smac/epm/partial_sparse_gaussian_process.py

```diff
@@ -0,0 +1,874 @@
+                        else:
+                            grad.append(t.detach().view(-1).cpu().double().clone().numpy())
+                    mll.zero_grad()
+                    return loss.item(), np.concatenate(grad)
+
+                theta_star = x0
+                f_opt_star = np.inf
+                for start_point in start_points:
+                    try:
+                        theta, f_opt, res_dict = optimize.fmin_l_bfgs_b(sci_opi_wrapper,
+                                                                        start_point,
+                                                                        args=(var_mll, property_dict, out_x_, out_y_),
+                                                                        bounds=bounds,
+                                                                        maxiter=50,
+                                                                        )
+                        if f_opt < f_opt_star:
+                            f_opt_star = f_opt
+                            theta_star = theta
+                    except Exception as e:
+                        self.logger.warning(f"An exception {e} occurs during the optimizaiton")
+
+                start_idx = 0
+                # modification on botorch.optim.numpy_converter.set_params_with_array
+                for p_name, attrs in property_dict.items():
+                    if p_name != "model.variational_strategy.inducing_points":
+                        # Construct the new tensor
+                        if len(attrs.shape) == 0:  # deal with scalar tensors
+                            start_idx = start_idx + 1
+                        else:
+                            start_idx = start_idx + np.prod(attrs.shape)
+                    else:
+                        end_idx = start_idx + np.prod(attrs.shape)
+                        inducing_points = torch.tensor(
+                            theta_star[start_idx:end_idx], dtype=attrs.dtype, device=attrs.device
+                        ).view(*attrs.shape)
+                        break
+                # set inducing points for covariance module here
+                self.gp_model.initialize(**{'covar_module.inducing_points': inducing_points})
+        else:
+            self.hypers, self.property_dict, _ = module_to_array(module=self.gp)
+
+        self.is_trained = True
+        return self
+
+    def _get_gp(self,
+                in_x: typing.Optional[np.ndarray] = None,
+                in_y: typing.Optional[np.ndarray] = None,
+                out_x: typing.Optional[np.ndarray] = None,
+                out_y: typing.Optional[np.ndarray] = None) -> typing.Optional[ExactMarginalLogLikelihood]:
+        """
+        Construction a new GP model based on the inputs
+        If both in and out are None: return an empty models
+        If only in_x and in_y are given: return a vanilla GP model
+        If in_x, in_y, out_x, out_y are given: return a partial sparse GP model.
```

Comment:
What is the use case of an empty model?
I would rather suggest to have in_x and in_y as required arguments and only out_x and out_y as optional.
Further, to be consistent with naming in other classes it would be good if you could rename the "x" variables to "in_X" and "out_X" respectively.

Reply:
smac.epm.base_gp.BaseModel requires to run 'self._get_gp()' in '__init__', I made all arguments optional to make it compatible with the super class